### ISSUE TITLE
Splicing a tuple into an indexed literal causes subsequent form to compile incorrectly

### ISSUE THREAD

**Issue by ianthehenry** [janet-lang/janet#1330]
Interesting bug where it seems like a `splice` is persisting after the expression it occurs in. This program:

```janet
(defn main [&]
  (def a [1 2 3])
  (def b [;a])
  (pp a))
```

Raises with `error: <function pp> called with 4 arguments, expected 1`. Commenting out the line with the splice causes it to work as expected.

Tested on `Janet 1.32.1-cc5beda0 macos/aarch64/clang`.

The following variations do not have the bug:

```janet
(defn main [&]
  (def a [1 2 3])
  (def b (tuple ;a))
  (pp a))

(defn main [&]
  (def a [1 2 3])
  (def b [;[1 2 3]])
  (pp a))

(defn main [&]
  (def a @[1 2 3])
  (def b [;a])
  (pp a))
```

Though this one also fails:

```janet
(defn main [&]
  (def a [1 2 3])
  (def b @[;a])
  (pp a))
```

**Comment by sogaiu** [comment]
Confirmed in `Janet 1.32.1-56f33f51 linux/x64/gcc`.

**Comment by sogaiu** [comment]
Janet 1.20.0-40ae2e81 does not exhibit the issue here.

**Comment by bakpakin** [comment]
Pushed a fix in a9176a77e6a93a5e10b5c069e9b1c88d4b25e40e - the issue seems to be that certain bytecode instructions were being aggressively optimized out despite have side effects (clearing the call stack), making the next function call get extra arguments.

**Comment by sogaiu** [comment]
All of the reported examples work fine here now :+1:
